server/lorax_server/models/custom_modeling/flash_phi_modeling.py

```diff
@@ -0,0 +1,396 @@
+# Adapted from
+# https://huggingface.co/microsoft/phi-2/blob/main/modeling_phi.py
+# and
+# https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/phi_1_5.py
+#
+# Copyright 2023 Predibase.
+# Copyright 2023 The vLLM team.
+# Copyright (c) Microsoft Corporation.
+#
+# LICENSE: https://huggingface.co/microsoft/phi-2/blob/main/LICENSE
+
+import torch
+import torch.distributed
+
+from torch import nn
+from transformers.activations import ACT2FN
+from transformers.configuration_utils import PretrainedConfig
+from transformers.models.phi import PhiConfig
+from typing import Optional, List, Tuple
+
+# Flash attention imports
+import dropout_layer_norm
+
+from lorax_server.utils import flash_attn
+from lorax_server.utils import paged_attn
+from lorax_server.utils.layers import (
+    FastLayerNorm,
+    TensorParallelAdapterRowLinear,
+    TensorParallelRowLinear,
+    TensorParallelColumnLinear,
+    TensorParallelEmbedding,
+    TensorParallelMultiAdapterLinear,
+    PositionRotaryEmbedding,
+    TensorParallelHead,
+)
+from lorax_server.utils.lora import LM_HEAD, AdapterBatchData
+
+
+ATTN_WQKV = "mixer.Wqkv"
+ATTN_OUT_PROJ = "mixer.out_proj"
+MLP_FC1 = "mlp.fc1"
+MLP_FC2 = "mlp.fc2"
+        
+
+def load_attention(config, prefix, weights, layer_id, head_dim, n_head, n_head_kv):
+    op_size = head_dim * (n_head + 2 * n_head_kv)
+    base_layer = load_attention_multi(config, prefix, weights, head_dim, n_head, n_head_kv)
+    return TensorParallelMultiAdapterLinear.load(
+        base_layer, layer_id, [ATTN_WQKV], sizes=[op_size], process_group=weights.process_group
+    )
+
+
+def load_attention_multi(config, prefix, weights, head_dim, n_head, n_head_kv):
+    return TensorParallelColumnLinear.load_multi(
+        config,
+        prefixes=[
+            (f"{prefix}.Wqkv", (0, head_dim * n_head)),
+            (f"{prefix}.Wqkv", (head_dim * n_head, head_dim * n_head_kv)),
+            (f"{prefix}.Wqkv", ((head_dim * n_head) + (head_dim * n_head_kv), head_dim * n_head_kv)),
+        ],
+        dim=0,
+        weights=weights,
+        bias=True,
+    )
+
+
+class FlashPhiAttention(torch.nn.Module):
+    def __init__(
+        self,
+        prefix: str,
+        config,
+        weights,
+        layer_id: int,
+    ):
+        super().__init__()
+        self.num_heads = config.n_head
+        self.hidden_size = config.n_embd
+        self.head_size = self.hidden_size // self.num_heads
```

Comment:
Is it intentionally to use `self.num_heads` like this here but then update the value of `self.num_heads` by the number of process groups in line 110?

Reply:
Yes, when creating the modules we need the `num_heads` value pre-splitting. Then later on we use `num_heads` post-split.

Reply:
Added comment.